Stratis.Bitcoin.Features.Miner/PosMinting.cs

```diff
@@ -323,22 +324,25 @@ private void CheckStake(ContextInformation context, ChainedBlock pindexPrev)
             }
 
             // Verify hash target and signature of coinstake tx.
-            BlockStake prevBlockStake = this.stakeChain.Get(pindexPrev.HashBlock);
+            BlockStake prevBlockStake = this.stakeChain.Get(prevBlock.HashBlock);
             if (prevBlockStake == null)
             {
                 this.logger.LogTrace("(-)[NO_PREV_STAKE]");
                 ConsensusErrors.PrevStakeNull.Throw();
             }
 
             context.SetStake();
-            this.posConsensusValidator.StakeValidator.CheckProofOfStake(context, pindexPrev, prevBlockStake, block.Transactions[1], block.Header.Bits.ToCompact());
+            this.posConsensusValidator.StakeValidator.CheckProofOfStake(context, prevBlock, prevBlockStake, block.Transactions[1], block.Header.Bits.ToCompact());
 
+            // The following is wrong, this should be REORG not SOLUTION FOUND
+            // but this would just narrow the race condition in case of reorg, not mitigate it.
+            // --------
             // Found a solution.
-            if (block.Header.HashPrevBlock != this.chain.Tip.HashBlock)
-            {
-                this.logger.LogTrace("(-)[SOLUTION_FOUND]");
-                return;
-            }
+            // if (block.Header.HashPrevBlock != chainTip.HashBlock)
```

Comment:
Its not reorg but rather the chain has advanced, I think it might still be needed.

Reply:
yes, the problem here is described in #488 
the check itself is useless - it only narrows the race condition window, which is dangerous thing, it is better to have wide race condition windows until the race is fixed

if the chain advanced after this IF was executed, then #488 crash happens, now it happens if the chain advances sooner, but if it advances at another point, we've got #500 crash

Reply:
The #500 is probably wrong to be a crash it should just be ignored especially when staking.
But yeah we should probably not use Tip any where but rather pass a ChainedBlock parameter instead (refactor).